DataDog/datadog-agent
# [system-probe] Use udp_destroy_sock probes to remove closed udp connections

### What does this PR do?

Use `udp_destroy_sock` kprobe/kretprobe to remove udp connections from the connections eBPF map. UDP destroy socket events are now sent down the perf buffer that was previously only used for tcp close events. 

In addition, there is a fix included for reading the remote address for `connect`-less receives, where a UDP `recvfrom` (in golang, `ReadFrom`) is done without connecting to the remote address/port. This causes the remote address in `struct sock*` passed in to `udp_recvmsg` not being set, and the tracer code ignoring this read. We now read the address from an out param to `udp_recvmsg`.

The above change also led to a restructuring of `read_conn_tuple`. It is not split into `read_conn_tuple` and `read_conn_tuple_uninit`. The latter has the meat of the code, with the change that it does not overwrite the values in the passed in `conn_tuple_t` unless they are `0`. `read_conn_tuple` is functionally the same as before.

### Motivation

UDP connections were only removed after a timeout prior to this change. On hosts with a lot of udp connection churn, this would lead to a large number of udp connections in memory. This change will at least remove the udp connections that are cleanly closed, including from the in-memory conntrack cache.

### Additional Notes

### Describe your test plan

1. Run the system-probe
2. Run `curl --unix-socket /var/run/sysprobe/sysprobe.sock http://unix/connections?client_id=1` in another console
3. In another console run `sudo socat PIPE UDP4-LISTEN:7,fork`
4. Run `nc -4 -u localhost 7` and enter any text; it should get echo'ed back. Leave this running
5. Run `curl --unix-socket /var/run/sysprobe/sysprobe.sock http://unix/connections?client_id=1 | jq '.conns[] | select(.type=="udp" and (.raddr.port==7 or .laddr.port==7))'`. Verify that you see two connections, one with `"type: incoming"` and the other with `"type": "outgoing"`. The `lastBytesSent` and `lastBytesReceived` should all be equal for both connections.
6. Now kill the `nc` command you left running in step (4.). Also kill the `socat` command. Wait 5-10 seconds
7. Run Run `curl --unix-socket /var/run/sysprobe/sysprobe.sock http://unix/connections?client_id=1 | jq '.conns[] | select(.type=="udp" and (.raddr.port==7 or .laddr.port==7))'`. You should not see any connections.
